### PR TITLE
ARROW-6558: [C++] Refactor Iterator to type erased handle

### PR DESCRIPTION
**Pull request by bkietz**
To resolve linkage issues with MSVC and the current abstract base class Iterator

@pitrou

### CODE REVIEW

**Comment by pitrou** [comment]
@kszucs Do you think the Ursabot build failures here are related?

File: cpp/src/arrow/table.h [hunk 1]
```diff
@@ -293,8 +291,12 @@ class ARROW_EXPORT TableBatchReader : public RecordBatchReader {
   void set_chunksize(int64_t chunksize);
 
  private:
-  class TableBatchReaderImpl;
-  std::unique_ptr<TableBatchReaderImpl> impl_;
```
Comment (pitrou) [inline cpp/src/arrow/table.h]: Why did you remove the pimpl? Is it just to remove one layer of indirection?

File: cpp/src/arrow/dataset/file_parquet_test.cc [hunk 2]
```diff
@@ -60,42 +60,43 @@ Status WriteRecordBatch(const RecordBatch& batch, FileWriter* writer) {
   return Status::OK();
 }
 
-Status WriteRecordBatchReader(RecordBatchReader* reader, FileWriter* writer) {
-  auto schema = reader->schema();
+Status WriteRecordBatchReader(RecordBatchReader&& reader, FileWriter* writer) {
```
Comment (pitrou) [inline cpp/src/arrow/dataset/file_parquet_test.cc]: Is there a reason for making these changes or is it just for aesthetics?

**Review by pitrou — COMMENTED** [review]
Thanks a lot for doing this. Here's a bunch of comments.

File: cpp/src/arrow/util/iterator.h [hunk 3]
```diff
@@ -28,84 +29,141 @@
...
 
       ARROW_RETURN_NOT_OK(visitor(std::move(value)));
     }
 
     return status;
   }
+
+  bool operator==(const Iterator& other) const { return ptr_ == other.ptr_; }
+
+  explicit operator bool() const { return ptr_ == NULLPTR; }
```
[the remark this answers is not in the document]
Comment (bkietz) [reply cpp/src/arrow/util/iterator.h]: That's a typo. Thanks

File: cpp/src/arrow/util/iterator.h [hunk 4]
```diff
@@ -28,84 +29,141 @@
...
+  return IteratorEnd(internal::type_constant<T>{});
+}
+
 /// \brief A generic Iterator that can return errors
 template <typename T>
 class Iterator {
  public:
-  virtual ~Iterator() = default;
+  /// \brief Iterator may be constructed from any type which has member function
+  /// Status HasNext::Next(T*);
```
[the remark this answers is not in the document]
Comment (bkietz) [reply cpp/src/arrow/util/iterator.h]: I can write it that way; using this notation is redundant when I just mentioned that it was a member function.

File: cpp/src/arrow/util/iterator.h [hunk 5]
```diff
@@ -28,84 +29,141 @@
...
 
-template <typename Fn, typename T = typename std::remove_pointer<
-                           internal::call_traits::argument_type<0, Fn>>::type>
-std::unique_ptr<FunctionIterator<Fn, T>> MakeFunctionIterator(Fn fn) {
-  return std::unique_ptr<FunctionIterator<Fn, T>>(
-      new FunctionIterator<Fn, T>(std::move(fn)));
+/// \brief Construct an Iterator which dereferences a (possibly smart) pointer
+/// to invoke its Next function
+template <typename Ptr,
+          typename Pointed = typename std::decay<decltype(*std::declval<Ptr>())>::type,
```
[the remark this answers is not in the document]
Comment (bkietz) [reply cpp/src/arrow/util/iterator.h]: `remove_pointer` would work for a raw pointer but this helper is mostly designed to wrap smart pointers (for which `remove_pointer` does nothing)

File: cpp/src/arrow/util/iterator.h [hunk 6]
```diff
@@ -28,84 +29,141 @@
...
+          typename Fn = decltype(std::mem_fun(&Pointed::Next)),
```
[the remark this answers is not in the document]
Comment (bkietz) [reply cpp/src/arrow/util/iterator.h]: I'll replace this with a custom trait

File: cpp/src/arrow/util/iterator.h [hunk 7]
```diff
@@ -28,84 +29,141 @@
...
+template <typename Fn, typename T = typename std::remove_pointer<
+                           internal::call_traits::argument_type<0, Fn>>::type>
+Iterator<T> MakeFunctionIterator(Fn fn) {
+  return Iterator<T>(FunctionIterator<Fn, T>(std::move(fn)));
+}
+
 template <typename T>
-std::unique_ptr<Iterator<T>> MakeEmptyIterator() {
-  return std::unique_ptr<EmptyIterator<T>>(new EmptyIterator<T>());
+Iterator<T> MakeEmptyIterator(Status s = Status::OK()) {
```
[the remark this answers is not in the document]
Comment (bkietz) [reply cpp/src/arrow/util/iterator.h]: I can remove the argument until we need it. This is slightly legacy: EmptyIterator's original purpose was to allow functions like `Iterator<unique_ptr<ScanTask>> NonStatusReturn()` to emit an error, but we decided that's an antipattern.

File: cpp/src/arrow/table.h (hunk 1 above)
Comment (bkietz) [reply cpp/src/arrow/table.h]: Yes, it's no longer necessary

File: cpp/src/arrow/util/iterator.h [hunk 8]
```diff
@@ -28,84 +29,141 @@
...
+
+/// \brief IteratorEnd returns a reserved value which indicates the end of iteration. By
+/// default this is NULLPTR since most iterators yield pointer types. Specialize/overload
+/// if different end semantics are required.
+template <typename T>
+static T IteratorEnd(internal::type_constant<T>) {
+  return T(NULLPTR);
+}
+
+/// Convenience function, overload the type_constant version above and not this one.
```
[the remark this answers is not in the document]
Comment (bkietz) [reply cpp/src/arrow/util/iterator.h]: The zero argument form cannot be partially specialized. Partial specialization is necessary if you need to define a generic `IteratorEnd`, for example for `optional`:

```c++
template <typename T>
optional<T> IteratorEnd<optional<T>>() {
  // signal end of iteration by yielding an empty optional
  return nullopt;
}
```

[the remark this answers is not in the document]
Comment (pitrou) [reply cpp/src/arrow/util/iterator.h]: Wow.

File: cpp/src/arrow/util/iterator.h [hunk 9]
```diff
@@ -28,84 +29,141 @@
...
  private:
-  Status status_;
+  Fn fn_;
 };
 
+/// \brief Construct an Iterator which invokes a callable on Next()
+template <typename Fn, typename T = typename std::remove_pointer<
+                           internal::call_traits::argument_type<0, Fn>>::type>
+Iterator<T> MakeFunctionIterator(Fn fn) {
+  return Iterator<T>(FunctionIterator<Fn, T>(std::move(fn)));
```
[the remark this answers is not in the document]
Comment (bkietz) [reply cpp/src/arrow/util/iterator.h]: It doesn't matter here: the two are equivalent *in the case of taking ownership*, since a new instance of Fn is constructed either way.

using forwarding:
```
template <typename T>
void append(vector<T>* vec, T&& t) {
  vec->push_back(std::forward<T>(t));
}

int main() {
  std::string str = "hello";
  append(&vec_of_string, str); // lvalue reference to str passed as argument
  append(&vec_of_string, std::move(str)); // rvalue reference to str passed as argument
}
```


using move:
```
template <typename T>
void append(vector<T>* vec, T t) {
  vec->push_back(std::move(t));
}

int main() {
  std::string str = "hello";
  append(&vec_of_string, str); // parameter is copy constructed from str
  append(&vec_of_string, std::move(str)); // parameter is move constructed from str
}
```

File: cpp/src/arrow/dataset/file_parquet_test.cc (hunk 2 above)
Comment (bkietz) [reply cpp/src/arrow/dataset/file_parquet_test.cc]: I was making it more clear from the function signature that the RecordBatchReader is consumed by WriteRecordBatchReader. Aesthetics, I suppose

File: cpp/src/arrow/util/iterator.h [hunk 10]
```diff
@@ -28,84 +29,141 @@
...
 
     return status;
   }
+
+  bool operator==(const Iterator& other) const { return ptr_ == other.ptr_; }
+
+  explicit operator bool() const { return ptr_ == NULLPTR; }
+
+ private:
+  template <typename HasNext>
```
[the remark this answers is not in the document]
Comment (bkietz) [reply cpp/src/arrow/util/iterator.h]: Will do

File: cpp/src/arrow/util/iterator.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (pitrou) [reply cpp/src/arrow/util/iterator.h]: Ah, I see!

File: cpp/src/arrow/util/iterator.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (pitrou) [reply cpp/src/arrow/util/iterator.h]: At worse, if someone wants to return an error deterministically, they can use `FunctionIterator`, no?

File: cpp/src/arrow/dataset/file_parquet_test.cc (hunk 2 above)
Comment (pitrou) [reply cpp/src/arrow/dataset/file_parquet_test.cc]: I don't know, I find it a bit confusing to be frank. We use `RecordBatchReader*` to point out that it's a non-const reference/pointer argument.

File: cpp/src/arrow/util/iterator.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (bkietz) [reply cpp/src/arrow/util/iterator.h]: I could also make this a static member function of a trait struct. That might be more clear:

```c++
template <typename T>
struct IterationTraits {
  static T End() { return NULLPTR; }
};

template <typename T>
struct IterationTraits<optional<T>> {
  static optional<T> End() { return nullopt; }
};
```

[the remark this answers is not in the document]
Comment (pitrou) [reply cpp/src/arrow/util/iterator.h]: That sounds good (though a bit more verbose for the caller).

File: cpp/src/arrow/dataset/file_parquet_test.cc (hunk 2 above)
Comment (bkietz) [reply cpp/src/arrow/dataset/file_parquet_test.cc]: alright, will do

File: cpp/src/arrow/util/iterator.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (bkietz) [reply cpp/src/arrow/util/iterator.h]: Done

File: cpp/src/arrow/util/iterator.h (hunk 10 above)
[the remark this answers is not in the document]
Comment (pitrou) [reply cpp/src/arrow/util/iterator.h]: Ah, sorry, I wasn't clear. I meant comment on the general design (why we're using type erasure) rather than details of member variables - these can stay of course :-)

**Comment by kszucs** [comment]
I have not seen the failures, everything seem to pass now though.

File: cpp/src/arrow/util/iterator.h (hunk 10 above)
[the remark this answers is not in the document]
Comment (pitrou) [reply cpp/src/arrow/util/iterator.h]: It can go in another PR though :-)

**Review by pitrou — APPROVED** [review]
+1

File: cpp/src/arrow/util/iterator.h (hunk 10 above)
[the remark this answers is not in the document]
Comment (bkietz) [reply cpp/src/arrow/util/iterator.h]: I'll add comments about this in #5221
